lib/web/automaticupgrades.go

```diff
@@ -0,0 +1,144 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package web
+
+import (
+	"context"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/mod/semver"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/integrations/kube-agent-updater/pkg/constants"
+	versionlib "github.com/gravitational/teleport/integrations/kube-agent-updater/pkg/version"
+	"github.com/gravitational/teleport/lib/automaticupgrades"
+)
+
+const defaultChannelTimeout = 5 * time.Second
+
+// automaticUpgrades implements a version server in the Teleport Proxy.
+// It is configured through the Teleport Proxy configuration and tells agent updaters
+// which version they should install.
+func (h *Handler) automaticUpgrades(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+	if h.cfg.AutomaticUpgradesChannels == nil {
+		return nil, trace.AccessDenied("This proxy is not configured to server automatic upgrades channels.")
+	}
+
+	// The request format is "<channel name>/{version,critical}"
+	// As <channel name> might contain "/" we have to split, pop the last part
+	// and re-construct the channel name.
+	channelAndType := p.ByName("request")
+
+	reqParts := strings.Split(strings.Trim(channelAndType, "/"), "/")
+	if len(reqParts) < 2 {
+		return nil, trace.BadParameter("path format should be /webapi/automaticupgrades/<channel>/{version,critical}")
+	}
+	requestType := reqParts[len(reqParts)-1]
+	channelName := strings.Join(reqParts[:len(reqParts)-1], "/")
+
+	if channelName == "" {
+		return nil, trace.BadParameter("a channel name is required")
+	}
+
+	// We check if the channel is configured
+	channel, ok := h.cfg.AutomaticUpgradesChannels[channelName]
+	if !ok {
+		return nil, trace.NotFound("channel %s not found", channelName)
+	}
+
+	// Finally, we treat the request based on its type
+	switch requestType {
+	case "version":
+		h.log.Debugf("Agent requesting version for channel %s", channelName)
+		return h.automaticUpgradesVersion(w, r, channel)
+	case "critical":
+		h.log.Debugf("Agent requesting criticality for channel %s", channelName)
+		return h.automaticUpgradesCritical(w, r, channel)
+	default:
+		return nil, trace.BadParameter("requestType path must end by 'version' or 'critical'")
+	}
+}
+
+// automaticUpgradesVersion handles version requests from upgraders
+func (h *Handler) automaticUpgradesVersion(w http.ResponseWriter, r *http.Request, channel *automaticupgrades.Channel) (interface{}, error) {
+	ctx, cancel := context.WithTimeout(r.Context(), defaultChannelTimeout)
+	defer cancel()
+
+	targetVersion, err := channel.GetVersion(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// We don't want to tell the updater to upgrade to a new major we don't support yet
```

Comment:
should we only consider major versions?
There are several cases where if auth server is behind agent version we can loose audit events because auth cannot/doesn't know how to unmarshal them from proto oneof.

We should always enforce proxies/auth to be the highest version (major.minor.patch) before upgrading agents. Sometimes we also use the same approach to release some features during minor upgrades

Reply:
The request from cloud folks was to explicitly block major versions to unlock their automatic update rollout. I think that running a higher bugfix version of the agents might be valid, at least, it should be compatible according to our compatibility guarantees.

Reply:
As discussed offline, I think this will bring more problems than solutions. The fact that agents can run multiple versions of the auth server (but always on the same major version) gives me the creeps.
There is nothing that guarantees that the auth server is not running 14.0.0 and the agents are not running 14.3.17 and this can be a big problem since we can have huge losses in audit logs. The same applies when agents report to prehog (discovery_service does this, machineid does too)

For the cloud it won't be a big problem but....


cc @rosstimothy do you have any concern about this or are you ok with it?

Reply:
This PR is not about the auth server version, it is about unblocking cloud automatic upgrades rollout because some tenants with mixed workloads block the whole cloud from upgrading.

I understand that you want to add new version restrictions but I don't think this is the right place.

Reply:
I suggested this because there is already a version check and version limiting. Doing the same for the auth server is trivial since each proxy has access to the auth server version in memory.

I'm not going to block the PR but I think we should have more guarantees that everything works when we decide to make upgrades.